Fixing exception for Table formatting type mismatch

bsipocz opened the pull request:
Rebased version of #3261 

Based on the comment https://github.com/astropy/astropy/pull/3261#issuecomment-132759746 I think that PR was good to go, just got merge conflicts.

fixes #2868

bsipocz commented:
this is a weird one, apparently locally it's seemingly randomly fails with ``E       AttributeError: 'Column' object has no attribute '_name'`` or passes without issues.

taldcroft commented:
@bsipocz - thanks for picking this up.  Please see comments:

https://github.com/astropy/astropy/pull/3261#issuecomment-132748187
https://github.com/astropy/astropy/pull/3261#r22369623
https://github.com/astropy/astropy/pull/3261#r22382401

taldcroft commented on astropy/table/column.py:
```diff
@@ -407,6 +410,42 @@ def name(self, val):
+            # test whether it formats without error exemplarily
+            self.pformat(max_lines=1)
+
+            if self.dtype.kind == 'O':
```
Even though this is formally correct to do the full pre-checking here, I think it is better to not have a very expensive operation.  One could end up formatting a billion lines just to set the`format` attribute, which is not a good thing.

bsipocz commented:
@taldcroft - I've mostly addressed your comments after redid the PR rebase fixes. I think most of the original comments were already dealt with, but please give it another review when you have time. 

I'll add a changelog, too once the CI passed.

taldcroft commented on astropy/table/pprint.py:
```diff
@@ -133,8 +133,8 @@ def _auto_format_func(format_, val):
                 # None of the possible string functions passed muster.
-                raise ValueError('Unable to parse format string {0}'
-                                 .format(format_))
+                raise TypeError('unable to parse format string {0} for its '
```
Keep as `ValueError`.   This is consistent with Python behavior.  The problem is not the *type* of the column but the *value* of the format string.

taldcroft commented on astropy/table/pprint.py:
```diff
@@ -431,29 +433,38 @@ def _pformat_col_iter(self, col, max_lines, show_name, show_unit, outs,
+                try:
+                    yield format_col_str(idx)
+                except TypeError:
+                    raise TypeError(
```
Catch `ValueError` and raise `ValueError`.

bsipocz commented:
@taldcroft - Done.

> Thanks for sticking with this!

No problem, this was a low hanging fruit for my occasional triaging of reducing the number of open issues.

taldcroft commented:
@bsipocz - Sorry, I had not noticed before this was milestoned for 2.0.2, but I think it should be 3.0.  The changelog could be something like:
```
New feature:
Improved exception handling and error messages when column ``format`` 
attribute is incorrect for the column type.

API change: 
When setting the column ``format`` attribute the value is now immediately validated. 
Previously one could set to any value and it was only checked when actually formatting the
column.
```

bsipocz commented:
Thanks @taldcroft! I meant to come back to this, but it slipped through...

taldcroft commented:
> I meant to come back to this, but it slipped through...

I'm very familiar with this phenomenon.
